paimon-core/src/main/java/org/apache/paimon/metrics/commit/CommitStats.java

```diff
@@ -0,0 +1,376 @@
+        this.changelogFilesCommitAppended = appendChangelogFiles.size();
+        this.changelogFilesCompacted = compactChangelogFiles.size();
+        this.numPartitionsWritten = numChangedPartitions(appendTableFiles, compactTableFiles);
+        this.numBucketsWritten = numChangedBuckets(appendTableFiles, compactTableFiles);
+        this.partitionsWritten = changedPartitions(appendTableFiles, compactTableFiles);
+        this.bucketsWritten = changedBuckets(appendTableFiles, compactTableFiles);
+        this.duration = commitDuration;
+        this.generatedSnapshots = generatedSnapshots;
+        this.attempts = attempts;
+    }
+
+    /** Return a map group by partition. */
+    @VisibleForTesting
+    protected static Map<BinaryRow, List<DataFileMeta>> groupByPartititon(
+            List<ManifestEntry> files) {
+        Map<BinaryRow, List<DataFileMeta>> groupByPartition = new LinkedHashMap<>();
+        for (ManifestEntry entry : files) {
+            groupByPartition
+                    .computeIfAbsent(entry.partition(), k -> new ArrayList<>())
+                    .add(entry.file());
+        }
+        return groupByPartition;
+    }
+
+    /** Return a map group by bucket. */
+    @VisibleForTesting
+    protected static Map<Integer, List<DataFileMeta>> groupByBucket(List<ManifestEntry> files) {
+        Map<Integer, List<DataFileMeta>> groupByBucket = new LinkedHashMap<>();
+        for (ManifestEntry entry : files) {
+            groupByBucket.computeIfAbsent(entry.bucket(), k -> new ArrayList<>()).add(entry.file());
+        }
+        return groupByBucket;
+    }
+
+    @VisibleForTesting
+    protected static long numChangedPartitions(List<ManifestEntry>... changes) {
+        return Arrays.stream(changes)
+                .flatMap(Collection::stream)
+                .map(ManifestEntry::partition)
+                .distinct()
+                .collect(Collectors.toList())
+                .size();
+    }
+
+    @VisibleForTesting
+    protected static long numChangedBuckets(List<ManifestEntry>... changes) {
+        return Arrays.stream(changes)
+                .flatMap(Collection::stream)
+                .map(ManifestEntry::bucket)
+                .distinct()
+                .collect(Collectors.toList())
+                .size();
+    }
+
+    @VisibleForTesting
+    protected static List<BinaryRow> changedPartitions(List<ManifestEntry>... changes) {
+        return Arrays.stream(changes)
+                .flatMap(Collection::stream)
+                .map(ManifestEntry::partition)
+                .distinct()
+                .collect(Collectors.toList());
+    }
+
+    @VisibleForTesting
+    protected static List<Integer> changedBuckets(List<ManifestEntry>... changes) {
+        return Arrays.stream(changes)
+                .flatMap(Collection::stream)
+                .map(ManifestEntry::bucket)
+                .distinct()
+                .collect(Collectors.toList());
+    }
```

Comment:
Same as above. Even with the same bucket number, if the partitions are different, they are considered to be different buckets.